src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs

```diff
@@ -493,28 +588,38 @@ private void HandleRegisterTopic(IActorRef actorRef)
         private void PutToRegistry(string key, IActorRef value)
         {
             var v = NextVersion();
-            if (!_registry.TryGetValue(_cluster.SelfAddress, out var bucket))
-                _registry.Add(_cluster.SelfAddress,
-                    new Bucket(_cluster.SelfAddress, v, ImmutableDictionary<string, ValueHolder>.Empty.Add(key, new ValueHolder(v, value))));
-            else
-                _registry[_cluster.SelfAddress] = new Bucket(bucket.Owner, v, bucket.Content.SetItem(key, new ValueHolder(v, value)));
+            var newBucket = _registry.TryGetValue(_cluster.SelfAddress, out var bucket)
+                ? new Bucket(bucket.Owner, v, bucket.Content.SetItem(key, new ValueHolder(v, value)))
+                : new Bucket(_cluster.SelfAddress, v, ImmutableDictionary<string, ValueHolder>.Empty.Add(key, new ValueHolder(v, value)));
+            
+            _registry[_cluster.SelfAddress] = newBucket;
+
+            // NOTE: We can't process this immediately here because message publishing needs to be done
+            //       AFTER we send `SubscribeAck`
+            _newlyAddedKeys.Add(key);
         }
 
-        private void IgnoreOrSendToDeadLetters(IWrappedMessage message)
+        private void IgnoreOrSendToDeadLetters(object message, IActorRef sender)
         {
+            if (message is PublishWithAck needAck)
+            {
+                sender.Tell(new PublishFailed(needAck, PublishFailReason.Timeout));
```

Comment:
Do we want to do an early return here and skip logging the dead letter or do we want to do both? The way we have it does both and I actually think that's ok - since the `DeadLetter` monitoring will usually get universal attention.